src/firestore/delete.ts

```diff
@@ -0,0 +1,486 @@
+              },
+            },
+          },
+          {
+            fieldFilter: {
+              field: {
+                fieldPath: "__name__",
+              },
+              op: "LESS_THAN",
+              value: {
+                referenceValue: endAt,
+              },
+            },
+          },
+        ],
+      },
+    };
+
+    const query: any = {
+      structuredQuery: {
+        where: where,
+        limit: batchSize,
+        from: [
+          {
+            allDescendants: allDescendants,
+          },
+        ],
+        select: {
+          fields: [{ fieldPath: "__name__" }],
+        },
+        orderBy: [{ field: { fieldPath: "__name__" } }],
+      },
+    };
+
+    if (startAfter) {
+      query.structuredQuery.startAt = {
+        values: [{ referenceValue: startAfter }],
+        before: false,
+      };
+    }
+
+    return query;
+  }
+
+  /**
+   * Construct a StructuredQuery to find descendant documents of a document.
+   * The document itself will not be included
+   * among the results.
+   *
+   * See:
+   * https://firebase.google.com/docs/firestore/reference/rest/v1beta1/StructuredQuery
+   *
+   * @param {boolean} allDescendants true if subcollections should be included.
+   * @param {number} batchSize maximum number of documents to target (limit).
+   * @param {string=} startAfter document name to start after (optional).
+   * @return {object} a StructuredQuery.
+   */
+  private _docDescendantsQuery(allDescendants: boolean, batchSize: number, startAfter?: string) {
+    const query: any = {
+      structuredQuery: {
+        limit: batchSize,
+        from: [
+          {
+            allDescendants: allDescendants,
+          },
+        ],
+        select: {
+          fields: [{ fieldPath: "__name__" }],
+        },
+        orderBy: [{ field: { fieldPath: "__name__" } }],
+      },
+    };
+
+    if (startAfter) {
+      query.structuredQuery.startAt = {
+        values: [{ referenceValue: startAfter }],
+        before: false,
+      };
+    }
+
+    return query;
+  }
+
+  /**
+   * Query for a batch of 'descendants' of a given path.
+   *
+   * For document format see:
+   * https://firebase.google.com/docs/firestore/reference/rest/v1beta1/Document
+   *
+   * @param {boolean} allDescendants true if subcollections should be included,
+   * @param {number} batchSize the maximum size of the batch.
+   * @param {string=} startAfter the name of the document to start after (optional).
+   * @return {Promise<object[]>} a promise for an array of documents.
+   */
+  private _getDescendantBatch(allDescendants: boolean, batchSize: number, startAfter?: string) {
+    const url = this.parent + ":runQuery";
+    let body;
+    if (this.isDocumentPath) {
+      body = this._docDescendantsQuery(allDescendants, batchSize, startAfter);
+    } else {
+      body = this._collectionDescendantsQuery(allDescendants, batchSize, startAfter);
+    }
+
+    return api
+      .request("POST", "/v1beta1/" + url, {
```

Comment:
When you improve this code, consider using `apiv2.Client`'s `post` (or `request`, if a lot of custom handling is required) method!

Reply:
I'll do this one in the other PR